### PR TITLE
[14.0][FIX] barcodes_generator_product: #516

### PR DESCRIPTION
**Pull request by ilyasProgrammer**
fix https://github.com/OCA/stock-logistics-barcode/issues/516

### CODE REVIEW

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi. Thanks for the fix. Some remarks :   
- you should call generate barcode only if it is an automated rule.
- could you add test ?
- the current design (overload onchange) doesnt work on import. I think implement on create or write is a better design.
- could you implement the patch in the abstract module ? So it will work for all modules. (Lot / partner / ....)


Thanks !

**Comment by ilyasProgrammer** [comment]
@legalsylvain Thanks for feedback. Please check new version.

File: barcodes_generator_abstract/models/barcode_generate_mixin.py [hunk 1]
```diff
@@ -29,14 +29,25 @@ class BarcodeGenerateMixin(models.AbstractModel):
...
 
+    def write(self, vals):
+        res = super(BarcodeGenerateMixin, self).write(vals)
+        if vals.get("barcode_rule_id"):
+            rule = self.env["barcode.rule"].browse(vals["barcode_rule_id"])
+            if rule.generate_automate and rule.generate_type == "sequence":
+                for rec in self:
+                    if not rec.barcode:
+                        rec.generate_base()
+                        rec.generate_barcode()
```
Comment (legalsylvain) [inline barcodes_generator_abstract/models/barcode_generate_mixin.py]:
```suggestion
                    if not rec.barcode_base:
                        rec.generate_base()
                     if not rec.barcode:
                        rec.generate_barcode()
```

(if base is defined, regenerate it is not required, don't you think ?)

File: barcodes_generator_abstract/models/barcode_generate_mixin.py [hunk 2]
```diff
@@ -29,14 +29,25 @@ class BarcodeGenerateMixin(models.AbstractModel):
         related="barcode_rule_id.generate_type",
     )
 
+    def write(self, vals):
+        res = super(BarcodeGenerateMixin, self).write(vals)
```
Comment (legalsylvain) [inline barcodes_generator_abstract/models/barcode_generate_mixin.py]:
```suggestion
        res = super().write(vals)
```

nitpicking : lighter syntax.

**Review by legalsylvain — COMMENTED** [review]
Thanks a lot ! 
I think the new design is more robust, and covers more cases.

2 minors remarks inline.

**Review by legalsylvain — APPROVED** [review]
Thanks a lot ! 

/ocabot merge patch

Note : I migrated recently the modules to 16.0 version [here](https://github.com/OCA/stock-logistics-barcode/pull/450). if you can port this patch, it would be great!

**Comment by francesco-ooops** [comment]
@legalsylvain shall we add you as maintainer for _abstract too?

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

**Comment by legalsylvain** [comment]
I'm defined as the maintainers here : 
- https://github.com/OCA/stock-logistics-barcode/blob/16.0/barcodes_generator_abstract/__manifest__.py#L13
- https://github.com/OCA/stock-logistics-barcode/blob/16.0/barcodes_generator_product/__manifest__.py#L17

so no need to add it.

Don't understand why bot doesn't allow me to merge. 
Waiting for a repo maintainer.

**Comment by francesco-ooops** [comment]
@legalsylvain but you're not for v14 ;)

**Pull request closed without merging**